### ISSUE TITLE
GUI elements: sizing/visibility difficulties at small sizes

### ISSUE THREAD

**Issue by thumpclang** [plugdata-team/plugdata#1402]
plugdata v0.8.3 standalone
M2 MacBook Pro (running macOS 14.2.1)

This most recent version of plugdata seems to be stable enough to handle my performance patch’s _massive_ GUI, which is an exciting development! However I’m observing sizing differences between vanilla Pd and plugdata, which are particularly noticeable with extremely tight/small GUI layouts. I understand that one cannot expect pixel perfect cross-platform consistency, but indeed the closer plugdata can get to vanilla sizing the more folks will be able to easily jump on board without having to redo their layouts. I put together a small test patch to show some of the more significant differences:

(Pd 0.54-1 on the left, plugdata 0.8.3 on the right)

https://github.com/plugdata-team/plugdata/assets/88018389/f0bf6316-96c6-4cc8-b857-0673764a1f2f

[small_gui_test.pd.zip](https://github.com/plugdata-team/plugdata/files/14018850/small_gui_test.pd.zip)

Before reporting my observations, it’s worth noting that vanilla seems to be able to set a smaller minimum size for GUI elements than plugdata does; in plugdata, if you try to resize to any value smaller than 12 (typed into the Inspector/Dimensions/Size field) it will clip to a minimum of 12. Interestingly, however, you can get down to a minimum of 11 using a GUI element’s resize handles. In vanilla (which is where I made my test patch), one can set elements to be as small as 8. As you can see from my test patch, plugdata is able to render these size 8/10 elements, but there are some issues. Starting with the most serious:

---------------------------------------

### interactive/center part of radio/toggle/bang too small or not visible
Particularly with vradio/hradio, the center square of an active cell is barely filled in, and by size 8 is completely invisible. Toggle (the center cross) and bang (the center circle) also have much smaller active areas than their vanilla counterparts, which makes it difficult to see activity & interact without hiding the element behind one’s mouse cursor. It’d be great if the central “active” area of each GUI were proportionally much bigger at these small sizes, as with the vanilla versions.

### radio/slider sizing
It looks like plugdata renders both vradio & hradio bigger than vanilla versions, leading to overlaps in the longer dimension. With the sliders, curiously hslider looks quite consistent with vanilla whereas vslider renders with too much width (while height is OK).

### fonts
All comment fonts seem to render shifted to the right, perhaps by 5-10 pixels, and also just a couple of pixels too far down (the horizontal/vertical hash marks I put in this test patch are comments with “|” or “—“ to show alignment), perhaps due to the font appearing a bit larger overall? Labels seem to be the opposite: all sizes render smaller than in vanilla, but still with a noticeable x shift to the right. num2 seems to render its font size in relation to Dimension regardless of Label Height, whereas vanilla increases the number font in relation to Label Height.

### canvas font size?
Is there no current way to set canvas/patch font size in plugdata? (if there is I can’t find it) I originally made my test patch in vanilla with canvas font size 10, then opening in plugdata it seems to render at size 12, so I had to go back & redo the layout at 12 in vanilla for consistency of the test. It would be good to be able to adjust this since vanilla patches have been made using various font settings, and using a different one than original leads to overlaps & other problems.

---------------------------------------

I’m sure many of these are not easy fixes, and I may not be a typical user, but I thought I’d at least report this in case it might open the door for more vanilla users to be able to join the party. Thanks for your amazing work on this project!

**Comment by 60-hz** [comment]
I agree that the size of objects could be a little closer to vanilla pd, so that many help files boxes don't overlap and readability would improve.
the issue is the size of inlets / outlets relative to boxes... making them half circles take a lot of space and the text is stuck in small size between them.

**Comment by timothyschoen** [comment]
> plugdata v0.8.3 standalone M2 MacBook Pro (running macOS 14.2.1)
> 
> This most recent version of plugdata seems to be stable enough to handle my performance patch’s _massive_ GUI, which is an exciting development! However I’m observing sizing differences between vanilla Pd and plugdata, which are particularly noticeable with extremely tight/small GUI layouts. I understand that one cannot expect pixel perfect cross-platform consistency, but indeed the closer plugdata can get to vanilla sizing the more folks will be able to easily jump on board without having to redo their layouts. I put together a small test patch to show some of the more significant differences:
> 
> (Pd 0.54-1 on the left, plugdata 0.8.3 on the right)
> small_gui_test_video.mov
> 
> [small_gui_test.pd.zip](https://github.com/plugdata-team/plugdata/files/14018850/small_gui_test.pd.zip)
> 
> Before reporting my observations, it’s worth noting that vanilla seems to be able to set a smaller minimum size for GUI elements than plugdata does; in plugdata, if you try to resize to any value smaller than 12 (typed into the Inspector/Dimensions/Size field) it will clip to a minimum of 12. Interestingly, however, you can get down to a minimum of 11 using a GUI element’s resize handles. In vanilla (which is where I made my test patch), one can set elements to be as small as 8. As you can see from my test patch, plugdata is able to render these size 8/10 elements, but there are some issues. Starting with the most serious:
> ### interactive/center part of radio/toggle/bang too small or not visible
> 
> Particularly with vradio/hradio, the center square of an active cell is barely filled in, and by size 8 is completely invisible. Toggle (the center cross) and bang (the center circle) also have much smaller active areas than their vanilla counterparts, which makes it difficult to see activity & interact without hiding the element behind one’s mouse cursor. It’d be great if the central “active” area of each GUI were proportionally much bigger at these small sizes, as with the vanilla versions.
> ### radio/slider sizing
> 
> It looks like plugdata renders both vradio & hradio bigger than vanilla versions, leading to overlaps in the longer dimension. With the sliders, curiously hslider looks quite consistent with vanilla whereas vslider renders with too much width (while height is OK).
> ### fonts
> 
> All comment fonts seem to render shifted to the right, perhaps by 5-10 pixels, and also just a couple of pixels too far down (the horizontal/vertical hash marks I put in this test patch are comments with “|” or “—“ to show alignment), perhaps due to the font appearing a bit larger overall? Labels seem to be the opposite: all sizes render smaller than in vanilla, but still with a noticeable x shift to the right. num2 seems to render its font size in relation to Dimension regardless of Label Height, whereas vanilla increases the number font in relation to Label Height.
> ### canvas font size?
> 
> Is there no current way to set canvas/patch font size in plugdata? (if there is I can’t find it) I originally made my test patch in vanilla with canvas font size 10, then opening in plugdata it seems to render at size 12, so I had to go back & redo the layout at 12 in vanilla for consistency of the test. It would be good to be able to adjust this since vanilla patches have been made using various font settings, and using a different one than original leads to overlaps & other problems.
> 
> I’m sure many of these are not easy fixes, and I may not be a typical user, but I thought I’d at least report this in case it might open the door for more vanilla users to be able to join the party. Thanks for your amazing work on this project!

Thanks for this feedback!

In terms of object size, plugdata is somewhere inbetween Max and pd-vanilla. Having a combo for height with 3 options, Max/plugdata/pd might be nice? I think I don't want to make this completely variable, since I can't possibly test all the options, and it seems like the kind of thing that can cause issues.

I'll also take a look at all the other size/offset issues, those should definitely get fixed.

**Comment by timothyschoen** [comment]
@60-hz There is a theming option for square iolets:

<img width="259" alt="Screenshot 2024-01-25 at 14 44 35" src="https://github.com/plugdata-team/plugdata/assets/44585538/14ce2c32-bad4-46b8-801b-da1adea86d3d">

I think I do like having larger hitboxes for iolets, they are very small in pd-vanilla, pretty difficult to click sometimes. The margins are copied from Max, but we could maybe add an option to disable that. That would make our pd-vanilla theme look more accurate than it looks right now:




<img width="449" alt="Screenshot 2024-01-25 at 14 46 52" src="https://github.com/plugdata-team/plugdata/assets/44585538/53690f66-3c69-441b-8d42-9d5a9804b8c6">

**Comment by 60-hz** [comment]
Thanks Tim, I am aware about the squared inlet themes, but it's more about the default theme is breaking a lot of help patches unfortunately because inlets not being on side of boxes I guess.

>I think I do like having larger hitboxes for iolets, they are very small in pd-vanilla, pretty difficult to click sometimes.
Yes I agree, recently the click zone in vanilla pd have expanded a lot, which is far better. But maybe there is a solution so that this extra space don't overlap / break some help patch?

**Comment by dromer** [comment]
I regularly run into this issue, where you simply cannot read a patch because boxes are completely overlapping.
This happening for downloaded patches is less of an issue that for vanilla help files.

The later should definitely be properly legible.

For the former something like an "auto-space" feature might be nice. Something that just pushes all objects away to a minimum margin (although this could have unforeseen consequences depending on the patching style). But this would be a completely separate feature request ;)

**Comment by 60-hz** [comment]
>I regularly run into this issue, where you simply cannot read a patch because boxes are completely overlapping.
This happening for downloaded patches is less of an issue that for vanilla help files.

Yes, I really think the only way to fix this is to have a closer vanilla size by default... if many helps are not clean and readable this is an serious issue for learning.

**Comment by timothyschoen** [comment]
Issues for sliders are now fixed!

**Comment by timothyschoen** [comment]
Also fixed a lot of problems for [hradio]/[vradio], should be more readable and more similar in size as vanilla

**Comment by thumpclang** [comment]
Awesome, thanks! Looking forward to checking these changes out.

Apologies if you've answered this elsewhere: is there any kind of auto-build system for folks to try out the latest changes in the develop branch? Or would one need to build their own binaries?

**Comment by timothyschoen** [comment]
Yes!

With a github account: https://github.com/plugdata-team/plugdata/actions -> click a build and scroll down
Without a github account: https://plugdata.org/download.html -> at the bottom of the page

**Comment by dromer** [comment]
> either I'm doing something dumb or there's a problem with the Mac builds

Please create a new issue or discussion with your problem. This topic is not a good place for it. :)

**Comment by thumpclang** [comment]
> Please create a new issue or discussion with your problem. This topic is not a good place for it. :)

@dromer Well, I admit to being inexperienced in this matter, however I have observed that some Github conversations become extremely difficult to follow when they're divided between multiple issues. I agree, super long threads that go way off topic are also hard to follow, so one has to draw the line _somewhere_ as to what exactly is off-topic.

In this case, I was asking about this crash in context of our current conversation as a sanity check, in case it might not be necessary to add noise to the Issues board (because I've never tried installing an auto-build before this, and it's totally possible that my inexperience is the problem). However, if @timothyschoen thinks it's easier for such comments to be spun off into new issues, I will certainly respect his preferences!

**Comment by timothyschoen** [comment]
Basically all of this is fixed!

- sliders, radios, etc. should have exactly the same size as in pd
- radio objects are now more readable at small sizes
- labels render more correctly, and try to squeeze the text into the bounds that pd would use for the label to get better positioning
- object height is now the same as in pd-vanilla, width is reduced a tiny bit too


Comment are still slightly offset, but it's pretty tricky to change that unfortunately.

**Comment by thumpclang** [comment]
I just tried the latest build... fantastic!! All GUI sizing looks really close to vanilla. Amazing work and a massive improvement, thank you! 🎉

The comment/text size seems like the only remaining major difference between plug & vanilla, and unfortunately beyond just general aesthetics this means there are still overlaps between paragraphs & objects in a lot of patches, including some vanilla documentation. I believe you that text sizing isn't an easy thing to adjust, however this still seems like a pretty significant fundamental difference; in vanilla, adjusting the font size per-canvas changes both comment size and also object rectangle size (to fit larger/smaller fonts), and nearly everyone generally lays out their patches for a specific font size (10, 12, 16, etc). Since plug already allows for custom fonts in the "Themes" tab, would it really not be feasible to add a font size option there which could apply to comments & object fonts?

Other random remaining plug-vanilla differences I noticed (all minor, but just for the sake of completeness):

- vanilla graphical arrays have a transparent background, some layouts rely on this (I think there's a vanilla doc example that uses overlapped arrays, can't find it ATM)
- size 8 toggle center "x" renders in plug like a dot because the arms of the "x" are too small to approach the toggle boundaries
- number2 font sizes still seem smaller than vanilla

No need to reopen this issue if you've moved on to other matters, the current improvements are already opening up a whole new world of usability for us longtime vanilla-using folks!

**Comment by timothyschoen** [comment]
The margin on the comment is now reduced, so it's a bit less bad now at least.

I'd like to make arrays transparent, but I think it would look messy when combined with the grid dots in edit mode. I'll consider it though. It would also solve another problem: the current contrast between array foreground and background is pretty low.

I've fixed toggles/bangs/sliders at very small sizes, and also made sure that the minimum size in plugdata is the same as in pd-vanilla.

I've also made the [number2] object font size bigger to match pd-vanilla.

Thanks for reporting all of this!

**Comment by thumpclang** [comment]
Awesome! Huge improvement again, especially with those tiny toggles. Again, thank you for these improvements!

If at any point you think the overall font size issue is worth revisiting, I'm happy to help with brainstorming a sensible way to do so (such as exposing adjustable font sizes in the Advanced menu, or some kind of "vanilla compat" font mode, etc). If not, I promise I'll stop blathering on about it. 😂 (I'm just genuinely excited by the prospect of a seamless transition away from the oppressive limitations of Tcl/Tk!)
